### PR TITLE
Use PathBuffer in off-policy algos

### LINKED ISSUE #1337: Move all off-policy algorithms to PathBuffer

### PR DESCRIPTION
**Pull request by maliesa96**
Fixes #1337. Modified `OffPolicyVectorizedSampler` and off-policy algos to use `PathBuffer`, and updated the tests, examples, and benchmarks to reflect this.

I also ran ddpg benchmarks with the changes to make sure there are no regressions:
https://tensorboard.dev/experiment/zF8t5K8aRQaozMaOZniDCw/

### CODE REVIEW

**Comment by avnishn** [comment]
Thanks for this!

**Comment by maliesa96** [comment]
@ryanjulian @avnishn this is ready for another review. The bug fix to `VecEnvExecutor` caused other problems, so I'll make that a separate PR.

File: src/garage/tf/algos/ddpg.py [hunk 1]
```diff
@@ -322,15 +334,23 @@ def optimize_policy(self, itr, samples_data):
...
+        if isinstance(self.replay_buffer, PathBuffer):
+            transitions = self.replay_buffer.sample_transitions(
+                self.buffer_batch_size)
+
+            observations = transitions['observations']
+            rewards = transitions['rewards']
+            actions = transitions['actions']
+            next_observations = transitions['next_observations']
+            terminals = transitions['terminals']
+        else:
```
Comment (ryanjulian) [inline src/garage/tf/algos/ddpg.py:346]: please leave TODO about removing this with an issue number from GitHub

File: src/garage/tf/algos/td3.py [hunk 2]
```diff
@@ -271,15 +272,23 @@ def optimize_policy(self, itr, samples_data):
...
+        if isinstance(self.replay_buffer, PathBuffer):
+            transitions = self.replay_buffer.sample_transitions(
+                self.buffer_batch_size)
+
+            observations = transitions['observations']
+            rewards = transitions['rewards']
+            actions = transitions['actions']
+            next_observations = transitions['next_observations']
+            terminals = transitions['terminals']
+        else:
```
Comment (ryanjulian) [inline src/garage/tf/algos/td3.py:284]: TODO

File: src/garage/torch/algos/ddpg.py [hunk 3]
```diff
@@ -155,10 +156,26 @@ def train_once(self, itr, paths):
...
+            if itr % self.steps_per_epoch == 0 and self._buffer_prefilled:
+                avg_success_rate = np.mean(self._success_history)
+
         for _ in range(self.n_train_steps):
             if self._buffer_prefilled:
-                samples = self.replay_buffer.sample(self.buffer_batch_size)
+                if isinstance(self.replay_buffer, PathBuffer):
+                    samples = self.replay_buffer.sample_transitions(
+                        self.buffer_batch_size)
+                else:
```
Comment (ryanjulian) [inline src/garage/torch/algos/ddpg.py:177]: TODO

File: src/garage/torch/algos/ddpg.py [hunk 4]
```diff
@@ -213,14 +229,19 @@ def optimize_policy(self, itr, samples_data):
...
-        rewards = rewards.reshape(-1, 1)
-        terminals = terminals.reshape(-1, 1)
+
+        if isinstance(self.replay_buffer, PathBuffer):
+            observations = transitions['observations']
+            rewards = transitions['rewards']
+            actions = transitions['actions']
+            next_observations = transitions['next_observations']
+            terminals = transitions['terminals']
+        else:
```
Comment (ryanjulian) [inline src/garage/torch/algos/ddpg.py:239]: TODO

**Comment by ryanjulian** [comment]
So your diff coverage is low because now a lot of the non-PathBuffer branches are uncovered.

You could remove the non-PathBuffer implementations, but I think that would require closing #1338.

Let me know when this is ready and we can force-push it.

**Comment by maliesa96** [comment]
@mergifyio rebase

**Comment by maliesa96** [comment]
@ryanjulian yeah I think #1338 should be addressed first. This is ready, though it looks like something in #1436 is causing the CI to fail.

**Comment by ryanjulian** [comment]
I'm cool with either order.

**Comment by ryanjulian** [comment]
@mergifyio rebase

**Pull request merged**